### DISCUSSION TITLE
sources_to_targets when there is an error for some points

### DISCUSSION THREAD

**Discussion by mertbakir** [valhalla/valhalla#3595] [Q&A] [answered]
Hi, I want to get distance between points using `sources_to_targets` endpoint but I know some of the points are not valid. 
So, I'm expecting nulls/errors for some values but I think I should get the values for the valid points in the matrix.

Example: Let's say I have 5 points, and sources and targets are the same 5 points. 2nd coordinate points to the ocean.
What I'm expecting: (take x as a valid metric, and -1 is error)

```
distance_matrix = 
[ x -1 x x x
 -1 -1 -1 -1 -1
 x -1 x x x
 x -1 x x x
 x -1 x x x]
```

What I get:
```
{"error_code":154,"error":"Path distance exceeds the max distance limit","status_code":400,"status":"Bad Request"
```

Is it possible to have a configuration to get the results for the valid points?

**Comment by kevinkreiser** [comment, marked as answer]
this error is telling you that some combination of the input locations are so far apart that they exceed the service limit you have configured. you can modify this configuration limit by editing your `valhalla.json` config directly or using the correct options when using `valhalla_build_config`. the ones you are looking for are here: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config#L177

to your other point about some of them not being possible. i forget if we made the code robust to that, i swear someone else asked for this feature somewhat recently and i dont remember us updating the code for us to allow failures of some pairs. someone correct me if im remembering wrong. i guess that means its not currently possible to get partial matrix results.

**Reply by mertbakir** [reply]
Thanks for the quick response. It would be really useful if we could've get partial results.

**Reply by kevinkreiser** [reply]
It's doable we could just not run the pairs that exceed the limits. Can you write up a feature request for it. I won't be able to work on it in the short term but it's a nice little chunk of work for a rainy day

**Reply by dnesbitt61** [reply]
there is also [this PR](https://github.com/valhalla/valhalla/pull/3181) which perhaps is an easy merge. However, this only supports one to many or many to one (using TimeDistanceMatrix) types of sources_to_target calls.

**Reply by kevinkreiser** [reply]
thats what i was failing to remember!

**Reply by mertbakir** [reply]
not related to the topic but may I ask what is TimeDistanceBSSMatrix. I know TimeDistanceMatrix. @dnesbitt61

**Reply by mertbakir** [reply]
here is the feature request @kevinkreiser 
https://github.com/valhalla/valhalla/issues/3596

**Reply by dnesbitt61** [reply]
TimeDistanceMatrixBSS is for bike share. As I understand it combines pedestrian and bicycle using bike share stations as "transfer" points between modes.

**Reply by mertbakir** [reply]
thanks.

**Reply by mertbakir** [reply]
@kevinkreiser 

It seems the PR is merged now. Can we say valhalla now returns partial matrix when some of the points are missing? I get a result like this today with distance and time values are `None`, Is this because of the new changes?
```
{'distance': None, 'time': None, 'to_index': 64, 'from_index': 0}, {'distance': None, 'time': None, 'to_index': 65, 'from_index': 0}
```

**Reply by kevinkreiser** [reply]
Yep when using that matrix algorithm we now return partial results.

**Reply by ghost** [reply]
> Yep when using that matrix algorithm we now return partial results.

Is this true for many-to-many requests? I still see the same error on http://valhalla1.openstreetmap.de/sources_to_targets.

**Reply by nilsnolde** [reply]
Nope, not for the CostMatrix

**Comment by nilsnolde** [comment]
So, for anyone reading this in the future: this whole thread is really confusing. I'll try to unravel the facts:
The OP is talking about two things:
- "invalid points" (whatever that exactly means, I assume "invalid routes")
- getting HTTP error when one or more matrix pair(s) distance(s) are beyond the server-side configured `max_matrix_distance` limit

So what's currently happening:
- we put `null` in the response for results of pairs where a connection couldn't be found (within the limits we allow for the expansion)
- we fail requests where even a single pair's distance is beyond the limit

What Kevin was referring to in https://github.com/valhalla/valhalla/discussions/3595#discussioncomment-3017091 was the fact that TimeDistanceMatrix has a new-ish request option `matrix_locations` which allows Valhalla to return partial results, i.e. only the user-specified amount of destinations for each origin, but it doesn't handle a pair exceeding the `max_distance_matrix` gracefully (see https://github.com/valhalla/valhalla/issues/3898#issuecomment-1377637254 for a suggestion).
